frontend/app/utils/date-utils.ts

```diff
@@ -51,3 +54,73 @@ export function getLocalizedMonths(
     text: formatter.format(Date.UTC(0, i, 1)), // formatted month name
   }));
 }
+
+/**
+ * Validates whether a given string is in the format YYYY-MM-DD.
+ *
+ * @param date - The date string to be validated.
+ * @returns Returns `true` if the string is a valid date in the format YYYY-MM-DD, otherwise `false`.
+ *
+ * @example
+ * ```javascript
+ * const validDate = "2023-05-20";
+ * const invalidDate = "20-05-2023";
+ *
+ * console.log(isValidDateString(validDate)); // true
+ * console.log(isValidDateString(invalidDate)); // false
+ * ```
+ */
+export const isValidDateString = (dateString: string) => {
+  return isValid(parseISO(dateString));
+};
+
+/**
+ * Validates whether a given string is in past or today.
+ *
+ * @param date - The date string to be validated.
+ * @returns Returns `true` if the string is a valid date in the format YYYY-MM-DD and the date is in past or today's date, otherwise `false`.
+ *
+ * This function first validates the input date string using the `isValidDateString` function.
+ * If the date string is invalid, an error is thrown. Otherwise, it compare date
+ * object with the date string set to the start of the day in UTC (00:00:00.000Z).
+ *
+ * @example
+ * ```javascript
+ * const validDateInPaste = "2023-05-20";
+ * const invalidDateInFuture = "6078-05-20";
+ *
+ * console.log(isDateInPastOrToday(validDateInPaste)); // true
+ * console.log(isDateInPastOrToday(invalidDateInFuture)); // false
+ * ```
+ */
+export function isDateInPastOrToday(dateString: string): boolean {
+  if (!isValidDateString(dateString)) {
+    throw new Error('Invalid date format. Expected YYYY-MM-DD.');
+  }
+  const inputDate = new Date(dateString);
+
+  const maxAllowedDate = getStartOfDayInTimezone('Canada/Eastern');
+
+  if (inputDate <= maxAllowedDate) return true;
+  return false;
+}
+
+/**
+ * Parses a time zone or a date (number or string or Date) and returns a date in the UTC (00:00:00.000Z) format.
+ * @param timeZone - The time zone as string.
+ * @returns A date set to the start of the day in UTC (00:00:00.000Z).
+ */
+export function getStartOfDayInTimezone(timezone: string, date?: number | string | Date): Date {
+  return new Date(startOfDay(TZDate.tz(timezone, new Date(date ?? Date.now()))));
+}
+
+/**
+ * Parses a date (year, month, day) and returns string in the format "YYYY-MM-DD".
+ * @param year - The year as number.
+ * @param month - The month as number.
+ * @param day - The day as number.
+ * @returns A string in the format "YYYY-MM-DD" (year, month, day).
+ */
+export function toISODateString(year: number, month: number, day: number): string {
+  return `${padWithZero(year, 4)}-${padWithZero(month, 2)}-${padWithZero(day, 2)}`;
+}
```

Comment:
```suggestion
/**
 * Formats a date into an ISO8601 date string (YYYY-MM-DD).
 *
 * This function takes the year, month, and day as numerical inputs and
 * returns a string representing the date in the "YYYY-MM-DD" format.
 *
 * @param year - The year (e.g., 2023).
 * @param month - The month (1 for January, 12 for December).
 * @param day - The day of the month.
 * @returns An ISO 8601 date string in the format "YYYY-MM-DD".
 */
export function toISODateString(year: number, month: number, day: number): string {
  return `${padWithZero(year, 4)}-${padWithZero(month, 2)}-${padWithZero(day, 2)}`;
}
```

Also.. I'd like to note that if you give this function something silly like `toISODateString(123, 3.1415, -Infinity)`, it will happily spit out `0123-3.1415--Infinity`. 😏

Reply:
we are getting the date (year as number, month as number and day s number) returned by the date picker field from the form. This value need to be converted to string before all the validations are done on that string. Maybe we need to change the name of this function??

Reply:
I think the function name is fine.. it's generating an ISO date string, just as the name suggests.

I think some guards should be added that prevent wonky inputs, though. It should probably throw if the year, month, or day are outside of the normal range. This might be doable using the date-fns parse and isValid functions.

Reply:
updated